uitools/toolkitcpp.cmake

```diff
@@ -11,42 +11,134 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-include(${CMAKE_CURRENT_LIST_DIR}/common.cmake)
+function(setup_toolkit TARGET_NAME TOOLKIT_DIR)
+    set(TOOLKITUI_PATH ${TOOLKIT_DIR}/uitools)
+    set(CPP_PATH ${TOOLKITUI_PATH}/cpp/Esri/ArcGISRuntime/Toolkit)
+    set(REGISTER_PATH ${TOOLKITUI_PATH}/register/Esri/ArcGISRuntime/Toolkit)
 
-set(REGISTERPATH ${CMAKE_CURRENT_LIST_DIR}/register/Esri/ArcGISRuntime/Toolkit)
+    setup_common_vars(TOOLKITCOMMON_HEADERS TOOLKITCOMMON_SOURCES)
+    setup_register_vars(TOOLKITREGISTER_HEADERS TOOLKITREGISTER_SOURCES)
 
-set(TOOLKITREGISTER_SOURCES
-    ${REGISTERPATH}/register.cpp
-    ${REGISTERPATH}/internal/register_cpp.cpp)
+    target_include_directories(${TARGET_NAME} PRIVATE
+        ${TOOLKITUI_PATH}/cpp
+        ${TOOLKITUI_PATH}/register
+        ${CPP_PATH}
+        ${REGISTER_PATH})
 
-set(TOOLKIT_RESOURCES
-    ${CMAKE_CURRENT_LIST_DIR}/images/esri_arcgisruntime_toolkit_images.qrc
-    ${CMAKE_CURRENT_LIST_DIR}/import/Esri/ArcGISRuntime/Toolkit/esri_arcgisruntime_toolkit_view.qrc)
+    target_sources(${TARGET_NAME} PRIVATE
+        ${TOOLKITCOMMON_HEADERS}
+        ${TOOLKITCOMMON_SOURCES}
+        ${TOOLKITREGISTER_HEADERS}
+        ${TOOLKITREGISTER_SOURCES}
+        ${TOOLKITUI_PATH}/images/esri_arcgisruntime_toolkit_images.qrc
+        ${TOOLKITUI_PATH}/import/Esri/ArcGISRuntime/Toolkit/esri_arcgisruntime_toolkit_view.qrc)
 
-target_include_directories(${PROJECT_NAME} PRIVATE
-    ${CMAKE_CURRENT_LIST_DIR}/register
-    ${REGISTERPATH})
+    find_package(Qt6 COMPONENTS REQUIRED QuickControls2 WebView Svg QuickLayouts)
 
-target_sources(${PROJECT_NAME} PRIVATE
-    ${TOOLKITCOMMON_SOURCES}
-    ${TOOLKITREGISTER_SOURCES}
-    ${TOOLKIT_RESOURCES})
+    target_link_libraries(${TARGET_NAME} PRIVATE
+        Qt6::QuickControls2
+        Qt6::WebView
+        Qt6::Svg)
 
-find_package(Qt6 COMPONENTS REQUIRED QuickControls2 WebView Svg QuickLayouts)
+    if(IOS)
+        target_link_libraries(${TARGET_NAME} PRIVATE Qt6::qquicklayoutsplugin)
+    else()
+        target_link_libraries(${TARGET_NAME} PRIVATE Qt6::QuickLayouts)
+    endif()
 
-target_link_libraries(${PROJECT_NAME} PRIVATE
-    Qt6::QuickControls2
-    Qt6::WebView
-    Qt6::Svg)
+    list(APPEND QML_IMPORT_PATH ${TOOLKITUI_DIR}/import)
 
-if(IOS)
-    target_link_libraries(${PROJECT_NAME} PRIVATE Qt6::qquicklayoutsplugin)
-else()
-    target_link_libraries(${PROJECT_NAME} PRIVATE Qt6::QuickLayouts)
-endif()
+    set(QML_IMPORT_PATH ${QML_IMPORT_PATH} CACHE STRING "" FORCE)
```

Comment:
There may be a problem here. I believe that each time cmake is run the path will be appended again and this cache variable and will continue to grow.

It's also possible that the user could stomp on what we set if they follow [the documentation](https://doc.qt.io/qtcreator/creator-qml-modules-with-plugins.html#importing-qml-modules).

I'm curious to know more about this QML_IMPORT_PATH cache variable. The documentation makes it sound like its only for Qt Creator. Is that correct?

Reply:
Good point, originally i had thought this was needed but I think I was incorrect.  It looks like this may be handled in our `Esri::ArcGISRuntime::Toolkit::registerComponents` function that we require users to call. So it may be the case that I can just get rid of these two lines.

I am not sure if the QML_IMPORT_PATH variable is exclusively used by the Qt Creator IDE, but I know that it is just another avenue to tell the QML engine where to look for modules you are importing, and where it can find the relevant `qmldir` folder for that module.